src/Package/Impl/Sql/Publish/SProcGenerator.cs

```diff
@@ -0,0 +1,179 @@
+        /// </summary>
+        private string FillSprocInlineTemplate(string rFilePath, string sprocName) {
+            var sprocTemplateFile = rFilePath.ToSProcFilePath();
+            var sprocTemplate = GetSqlFileContent(sprocTemplateFile);
+
+            var rCode = GetRFileContent(rFilePath);
+            rCode = rCode.EndsWithOrdinal(Environment.NewLine) ? rCode : rCode + Environment.NewLine;
+            sprocTemplate = sprocTemplate.Replace(RCodeTemplate, Environment.NewLine + rCode);
+
+            var sqlQuery = GetSqlFileContent(rFilePath.ToQueryFilePath()).Trim();
+            return sprocTemplate.Replace(InputQueryTemplate, sqlQuery);
+        }
+
+        private string FillSprocTableTemplage(string rFilePath, string sprocName, string codeTableName) {
+            var sprocTemplateFile = rFilePath.ToSProcFilePath();
+            var sprocTemplate = GetSqlFileContent(sprocTemplateFile);
+
+            var format =
+@"BEGIN
+DECLARE @RCodeQuery NVARCHAR(max);
+DECLARE @RCode NVARCHAR(max);
+DECLARE @ParmDefinition NVARCHAR(max);
+
+SET @RCodeQuery = N'SELECT @RCodeOUT = RCode FROM [{0}] WHERE SProcName = ''{1}''';
+SET @ParmDefinition = N'@RCodeOUT NVARCHAR(max) OUTPUT';
+
+EXEC sp_executesql @RCodeQuery, @ParmDefinition, @RCodeOUT=@RCode OUTPUT;
+SELECT @RCode;
+";
+            var declarations = string.Format(CultureInfo.InvariantCulture, format, codeTableName, sprocName);
+            sprocTemplate = sprocTemplate.Replace("BEGIN", declarations);
+            sprocTemplate = sprocTemplate.Replace("N'_RCODE_'", "@RCode");
+
+            var sqlQuery = GetSqlFileContent(rFilePath.ToQueryFilePath()).Trim();
+            return sprocTemplate.Replace(InputQueryTemplate, sqlQuery);
+        }
+
+        private void CreateStoredProcedures(SqlSProcPublishSettings settings, EnvDTE.Project targetProject, string targetFolder, EnvDTE.ProjectItem targetProjectItem) {
+            foreach (var rFilePath in settings.Files) {
+                var sprocName = settings.SProcNames[rFilePath];
+                if (!string.IsNullOrEmpty(sprocName)) {
+
+                    string template;
+                    if (settings.CodePlacement == RCodePlacement.Inline) {
+                        template = FillSprocInlineTemplate(rFilePath, sprocName);
+                    } else {
+                        template = FillSprocTableTemplage(rFilePath, sprocName, settings.TableName);
+                    }
+                    if (!string.IsNullOrEmpty(template)) {
+                        var sprocFile = Path.ChangeExtension(Path.Combine(targetFolder, sprocName), ".sql");
+                        _fs.WriteAllText(sprocFile, template);
+                        targetProjectItem.ProjectItems.AddFromFile(sprocFile);
+                    }
+                }
+            }
+        }
+
+        private string GetRFileContent(string filePath) {
+            if (_fs.FileExists(filePath)) {
+                return _fs.ReadAllText(filePath).Replace("'", "''");
```

Comment:
Then `_fs.FileExists(filePath)` is not needed

---
In reply to: [72572665](https://github.com/Microsoft/RTVS/pull/2122#discussion_r72572665) [](ancestors = 72572665)